services/QuillLMS/app/controllers/concerns/growth_results_summary.rb

```diff
@@ -0,0 +1,151 @@
+
+  NOT_PRESENT = 'Not present'
+  NOT_CORRECT = 'Not correct'
+  FULLY_CORRECT = 'Fully correct'
+  PARTIALLY_CORRECT = 'Partially correct'
+
+  NO_PROFICIENCY = 'No proficiency'
+  PARTIAL_PROFICIENCY = 'Partial proficiency'
+  PROFICIENCY = 'Proficiency'
+  GAINED_PROFICIENCY = 'Gained proficiency'
+  MAINTAINED_PROFICIENCY = 'Maintained proficiency'
+
+  extend self
+
+  def growth_results_summary(current_user, pre_test_activity_id, post_test_activity_id, classroom_id)
+    @current_user = current_user
+    pre_test = Activity.find(pre_test_activity_id)
+    @skill_groups = pre_test.skill_groups
+    set_pre_test_activity_sessions_and_assigned_students(pre_test_activity_id, classroom_id)
+    set_post_test_activity_sessions_and_assigned_students(post_test_activity_id, classroom_id)
+    @skill_group_summaries = @skill_groups.map do |skill_group|
+      {
+        name: skill_group.name,
+        description: skill_group.description,
+        not_yet_proficient_in_pre_test_student_names: [],
+        not_yet_proficient_in_post_test_student_names: []
+      }
+    end
+
+    {
+      student_results: student_results,
+      skill_group_summaries: @skill_group_summaries
+    }
+  end
+
+  private def set_pre_test_activity_sessions_and_assigned_students(activity_id, classroom_id)
+    unit_ids = @current_user.units.joins("JOIN unit_activities ON unit_activities.activity_id = #{activity_id}")
+    classroom_units = ClassroomUnit.where(unit_id: unit_ids, classroom_id: classroom_id)
+    assigned_student_ids = classroom_units.map { |cu| cu.assigned_student_ids }.flatten.uniq
+    @pre_test_assigned_students = User.where(id: assigned_student_ids).sort_by { |u| u.last_name }
+    @pre_test_activity_sessions = ActivitySession.where(activity_id: activity_id, classroom_unit_id: classroom_units.ids, state: 'finished').order(completed_at: :desc).uniq { |activity_session| activity_session.user_id }
+  end
+
+  private def set_post_test_activity_sessions_and_assigned_students(activity_id, classroom_id)
+    unit_ids = @current_user.units.joins("JOIN unit_activities ON unit_activities.activity_id = #{activity_id}")
+    classroom_units = ClassroomUnit.where(unit_id: unit_ids, classroom_id: classroom_id)
+    assigned_student_ids = classroom_units.map { |cu| cu.assigned_student_ids }.flatten.uniq
+    @post_test_assigned_students = User.where(id: assigned_student_ids).sort_by { |u| u.last_name }
+    @post_test_activity_sessions = ActivitySession.where(activity_id: activity_id, classroom_unit_id: classroom_units.ids, state: 'finished').order(completed_at: :desc).uniq { |activity_session| activity_session.user_id }
+  end
+
+  private def student_results
+    @post_test_assigned_students.map do |assigned_student|
+      post_test_activity_session = @post_test_activity_sessions.find { |as| as.user_id == assigned_student.id }
+      pre_test_activity_session = @pre_test_activity_sessions.find { |as| as.user_id == assigned_student.id }
```

Comment:
These could be slightly faster to do look-ups on.  If, instead of storing `@post_test_activity_sessions` in an array, you could but it in a hash: `@post_test_activity_sessions = @post_test_activity_sessions.map { |session| [session.user_id, session] }.to_h` (passing an array of size-2 arrays to `to_h` gives you a hash that uses the first value as a key and the second value as the value for the hash).  That lets you do your lookups on `@post_test_activity_session[assigned_student.id]`.  That means you don't have to loop over the arrays for each record in `@post_test_assigned_students`, and thus it is slightly faster.  (Takes longer to write the original value, but makes it faster to look up specific sections of it.)